### PR TITLE
[clipboard][web] Add support for images

### PR DESCRIPTION
**Pull request by barthap**
# Why

- Resolves ENG-4215
- Implements support for copying/pasting images on web (iOS: #16391, Android: #16413)

# How

- This implementation uses `AsyncClipboard` API: [intro/guide](https://web.dev/async-clipboard/), [W3C spec](https://www.w3.org/TR/clipboard-apis/#async-clipboard-api), [MDN](https://developer.mozilla.org/en-US/docs/Web/API/Clipboard_API), [browser support status](https://caniuse.com/async-clipboard)
- This API is still experimental, several browsers behave a bit differently, e.g. when requesting permissions - and this is subject to changes
- **Huge issue on WebKit/Safari**: the Clipboard API cannot be used in an async function, but only directly as a consequence of user action (e.g. directly in `onClick` event) - [see issue here](https://bugs.webkit.org/show_bug.cgi?id=222262) - we could work around this, but this workaround would have a huge impact on user code. I'd rather publish a polyfill/guide as a GH gist or sth and document it's not working in Safari.
   This has impact on our existing `Clipboard.getStringAsync()` - it doesn't work either.
- In contrast to iOS/Android, the `hasImageAsync()` requires user to accept clipboard permission (e.g. Chrome displays a popup)
- Refactored file structure

# Test Plan

- Works in Chrome 98 (NCL)
- Does not work in Safari (see above)

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: docs/pages/versions/unversioned/sdk/clipboard.md [hunk 1]
```diff
@@ -30,8 +30,7 @@ export default function App() {
   const [copiedText, setCopiedText] = React.useState('');
 
   const copyToClipboard = () => {
-    /* @info */ Clipboard.setString('hello world');/* @end */
-
+    /* @info */ Clipboard.setString('hello world'); /* @end */
```
Comment (tsapeta) [inline docs/pages/versions/unversioned/sdk/clipboard.md:33]: Shouldn't we use `setStringAsync` in the examples already?

**Review by tsapeta — APPROVED** [review]

File: docs/pages/versions/unversioned/sdk/clipboard.md (hunk 1 above)
Comment (barthap) [reply docs/pages/versions/unversioned/sdk/clipboard.md:33]: Yes, you're right. I need to update all docs examples.

File: packages/expo-clipboard/src/web/WebExpoClipboard.ts [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    const textField = document.createElement('textarea');
+    textField.textContent = text;
+    document.body.appendChild(textField);
+    textField.select();
+    try {
+      document.execCommand('copy');
+      success = true;
+    } catch (e) {}
+    document.body.removeChild(textField);
+    return success;
```
Comment (bbarthec) [inline packages/expo-clipboard/src/web/WebExpoClipboard.ts:56]: 👀 This way is not a good option? 🤔 
It's just a suggestion, but I wonder whether we should we more strict about not using such variables `let success = false` and modyfying them later.
```suggestion
    try {
      document.execCommand('copy');
      return true
    } catch (e) {
      return false
    } finally {
      document.body.removeChild(textField);
    }
```

File: packages/expo-clipboard/src/web/WebExpoClipboard.ts [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+  async setStringAsync(text: string, _options: SetStringOptions): Promise<boolean> {
+    return this.setString(text);
+  },
+  async hasStringAsync(): Promise<boolean> {
+    return this.getStringAsync({}).then((text) => text.length > 0);
+  },
+  async getImageAsync(_options: GetImageOptions): Promise<ClipboardImage | null> {
+    if (!navigator.clipboard) {
+      throw new ClipboardUnavailableException();
+    }
```
Comment (bbarthec) [inline packages/expo-clipboard/src/web/WebExpoClipboard.ts:67]: You don't check for `navigator.clipboard` existence in `getStringAsync` function.

File: packages/expo-clipboard/src/web/WebExpoClipboard.ts [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+  async hasImageAsync(): Promise<boolean> {
+    if (!navigator.clipboard) {
+      throw new ClipboardUnavailableException();
+    }
+
+    try {
+      const clipboardItems = await navigator.clipboard.read();
+      return clipboardItems
+        .flatMap((item) => item.types)
+        .some((type) => type === 'image/png' || type === 'image/jpeg');
```
Comment (bbarthec) [inline packages/expo-clipboard/src/web/WebExpoClipboard.ts:125]: Can you share this logic with `findImageInClipboardAsync`?
I imagine you can have:
```ts
async function readImagesFromClipboard() {
  const clipboardItems = await navigator.clipboard.read();
  return return clipboardItems
    .filter(item => item.types.some(type => ['image/png', 'image/jpeg'].contains(type))
}

const images = readImagesFromClipboard();

// 1. hasImageAsync
return hasImagesAsync.length > 0

// 2. getImageAsync
async function readImageBlobFromClipboardItem(item: ClipboardItem?) {
  // first look for png
    if (item?.types.some((type) => type === 'image/png')) {
      return item.getType('image/png');
    }

    // alternatively, an image might be a jpeg
    // NOTE: Currently, this is not supported by browsers yet. They only support PNG now
    if (item?.types.some((type) => type === 'image/jpeg')) {
      return clipboardItem.getType('image/jpeg');
    }
    
    return null
}
const blob = await readImageBlobFromClipboardItem(images[0])

File: packages/expo-clipboard/src/web/WebUtils.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+ */
+export function blobToBase64Async(blob: Blob): Promise<string> {
+  return new Promise((resolve, _) => {
+    const reader = new FileReader();
+    reader.onloadend = () => resolve(reader.result as string);
+    reader.readAsDataURL(blob);
+  });
+}
+
+export function getImageSizeFromBlobAsync(blob: Blob): Promise<[width: number, height: number]> {
```
Comment (bbarthec) [inline packages/expo-clipboard/src/web/WebUtils.ts:41]: I don't see a reason for using tuple here 🤔 
```suggestion
export function getImageSizeFromBlobAsync(blob: Blob): Promise<{ width: number, height: number }> {
```

File: packages/expo-clipboard/src/web/WebUtils.ts [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+  return new Promise((resolve, _) => {
+    const blobUrl = URL.createObjectURL(blob);
+    const img = document.createElement('img');
+    img.src = blobUrl;
+    img.onload = function () {
+      resolve([img.width, img.height]);
```
Comment (bbarthec) [inline packages/expo-clipboard/src/web/WebUtils.ts:47]:
```suggestion
      resolve({ width: img.width, height: img.height });
```

**Review by bbarthec — COMMENTED** [review]
Looking very promising 👍 
Left a few comments 😉

File: packages/expo-clipboard/src/ExpoClipboard.web.ts [hunk 7]
```diff
@@ -1,43 +1 @@
...
-  async setStringAsync(text: string, _options: SetStringOptions): Promise<boolean> {
-    return this.setString(text);
-  },
-  async hasStringAsync(): Promise<boolean> {
-    return this.getStringAsync({}).then((text) => text.length > 0);
-  },
-  addClipboardListener(): void {},
-  removeClipboardListener(): void {},
-};
+export { default } from './web/WebExpoClipboard';
```
[the remark this answers is not in the document]
Comment (barthap) [reply packages/expo-clipboard/src/ExpoClipboard.web.ts:1]: To keep all web logic inside the `web/` directory. I don't have strong opinion on this, but I find it clean this way.

File: packages/expo-clipboard/src/web/WebExpoClipboard.ts (hunk 2 above)
Comment (barthap) [reply packages/expo-clipboard/src/web/WebExpoClipboard.ts:56]: Old code (not mine 😅 ), but a good catch 👍

File: packages/expo-clipboard/src/ExpoClipboard.web.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (bbarthec) [reply packages/expo-clipboard/src/ExpoClipboard.web.ts:1]: Ok, I like the idea of having a separate directory for a web implementation, but I guess we should drop the prefix `Web*` from the files contained in `web/` directory. 🤔

File: packages/expo-clipboard/src/web/WebExpoClipboard.ts [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+        // Internet Explorer
+        // @ts-ignore
+        text = window.clipboardData.getData('Text');
+      } catch (e) {
+        return Promise.reject(new Error('Unable to retrieve item from clipboard.'));
+      }
+    }
+    return text;
+  },
+  setString(text: string): boolean {
```
[the remark this answers is not in the document]
Comment (bbarthec) [reply packages/expo-clipboard/src/web/WebExpoClipboard.ts:45]: Ok, I see that's required then. Mark it with a TODO suggesting removal in a few SDK cycles and add some Linear task for keeping track of it.

[the remark this answers is not in the document]
Comment (barthap) [reply packages/expo-clipboard/src/web/WebExpoClipboard.ts:45]: Created ENG-4464

File: packages/expo-clipboard/src/ExpoClipboard.web.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (barthap) [reply packages/expo-clipboard/src/ExpoClipboard.web.ts]: I dropped the `Web*` prefix and renamed `web/WebExpoClipboard` to `web/ClipboardModule` so we have the same "native" entry point for each platform: `android/ClipboardModule.kt`, `ios/ClipboardModule.swift` and `web/ClipboardModule.ts`

**Pull request merged**